README.md
# KdbxWeb ![CI Checks](https://github.com/keeweb/kdbxweb/workflows/CI%20Checks/badge.svg) [![Coverage Status](https://coveralls.io/repos/github/keeweb/kdbxweb/badge.svg?branch=master)](https://coveralls.io/github/keeweb/kdbxweb?branch=master)

KdbxWeb is a high-performance javascript library for reading/writing KeePass v2 databases (kdbx) in node.js or browser.

## Features

- runs in browser or node.js
- no native addons
- fast encryption with WebCrypto
- total ≈130kB with dependencies
- full support of Kdbx features
- protected values are stored in memory XOR'ed
- conflict-free merge support
- high code coverage
- strict TypeScript

## Browser support

- modern browsers: Chrome, Firefox, Safari, Opera, Edge
- node.js

## Compatibility

Supported formats are Kdbx3 and Kdbx4, current KeePass file format. Old kdb files (for KeePass v1) are out of scope of this library.

## Kdbx4

Kdbx4 has introduced Argon2, a new hashing function. Due to complex calculations, you have to implement it manually and export to kdbxweb, if you want to support such files. Here's how:

```ts
kdbxweb.CryptoEngine.setArgon2Impl((password, salt,
    memory, iterations, length, parallelism, type, version
) => {
    // your implementation makes hash (Uint8Array, 'length' bytes)
    return Promise.resolve(hash);
});
```

You can find an implementation example in [tests](https://github.com/keeweb/kdbxweb/blob/master/test/test-support/argon2.ts).  

It's not compiled into the library because there's no universal way to provide a fast implementation, so it's up to you, to choose the best one.

## Usage

##### Loading

```ts
let credentials = new kdbxweb.Credentials(kdbxweb.ProtectedValue.fromString('demo'),
    keyFileArrayBuffer, challengeResponseFunction);
const db1 = await kdbxweb.Kdbx.load(dataAsArrayBuffer, credentials);
const db2 = await kdbxweb.Kdbx.loadXml(dataAsString, credentials);
```

##### Saving

```ts
const dataAsArrayBuffer = await db.save();
const xmlAsString = await db.saveXml();
```

You can also pretty-print XML:
```ts
const prettyPrintedXml = await db.saveXml(true);
```

##### File info
```ts
db.header
db.meta
```

See the corresponding type fields inside, they should be obvious.

##### Changing credentials
```ts
const db = await kdbxweb.Kdbx.load(data, credentials);
db.credentials.setPassword(kdbxweb.ProtectedValue.fromString('newPass'));
const randomKeyFile = await kdbxweb.Credentials.createRandomKeyFile();
db.credentials.setKeyFile(randomKeyFile);
await db.save();
```

##### Creation

```ts
let newDb = kdbxweb.Kdbx.create(credentials, 'My new db');
let group = newDb.createGroup(newDb.getDefaultGroup(), 'subgroup');
let entry = newDb.createEntry(group);
```

##### Maintenance

```ts
db.cleanup({
    historyRules: true,
    customIcons: true,
    binaries: true
});

// upgrade the db to latest version (currently KDBX4)
db.upgrade();

// downgrade to KDBX3
db.setVersion(3);

// set KDF to AES
db.setKdf(kdbxweb.Consts.KdfId.Aes);
```

##### Merge

Entries, groups and meta are consistent against merging in any direction with any state.  
Due to format limitations, p2p entry history merging and some non-critical fields in meta can produce phantom records or deletions, 
so correct entry history merging is supported only with one central replica. Items order is not guaranteed but the algorithm tries to preserve it.
```ts
let db = await kdbxweb.Kdbx.load(data, credentials); // load local db
// work with db
db.save(); // save local db
let editStateBeforeSave = db.getLocalEditState(); // save local editing state (serializable to JSON)
db.close(); // close local db
db = kdbxweb.Kdbx.load(data, credentials); // reopen it again
db.setLocalEditState(editStateBeforeSave); // assign edit state obtained before save
// work with db
let remoteDb = await kdbxweb.Kdbx.load(remoteData, credentials); // load remote db
db.merge(remoteDb); // merge remote into local
delete remoteDb; // don't use remoteDb anymore
let saved = await db.save(); // save local db
editStateBeforeSave = db.getLocalEditState(); // save local editing state again
let pushedOk = pushToUpstream(saved); // push db to upstream
if (pushedOk) {
    db.removeLocalEditState(); // clear local editing state
    editStateBeforeSave = null; // and discard it
}
```

##### Groups
```ts
let defaultGroup = db.getDefaultGroup();
let anotherGroup = db.getGroup(uuid);
let deepGroup = defaultGroup.groups[1].groups[2];
```

##### Group creation
```ts
let group = db.createGroup(db.getDefaultGroup(), 'New group');
let anotherGroup = db.createGroup(group, 'Subgroup');
```

##### Group deletion
```ts
db.remove(group);
```

##### Group move
```ts
db.move(group, toGroup);
db.move(group, toGroup, atIndex);
```

##### Recycle Bin
```ts
let recycleBin = db.getGroup(db.meta.recycleBinUuid);
if (!recycleBin) {
    db.createRecycleBin();
}
```

##### Recursive traverse
```ts
for (const entry of group.allEntries()) { /* ... */ }
for (const group of group.allGroups()) { /* ... */ }
for (const entryOrGroup of group.allGroupsAndEntries()) { /* ... */ }
```

##### Entries
```ts
let entry = db.getDefaultGroup().entries[0];
entry.fields.AccountNumber = '1234 5678';
entry.fields.Pin = kdbxweb.ProtectedValue.fromString('4321');
```

##### Entry creation
```ts
let entry = db.createEntry(group);
```

##### Entry modification
```ts
// push current state to history stack
entry.pushHistory();
// change something
entry.fgColor = '#ff0000';
// update entry modification and access time
entry.times.update();
// remove states from entry history
entry.removeHistory(index, count);
```
Important: don't modify history states directly, this will break merge.

##### Entry deletion
```ts
db.remove(entry);
```

##### Entry move
```ts
db.move(entry, toGroup);
```

If you're moving an entry from another file, this is called _import_:
```ts
db.importEntry(entry, toGroup, sourceFile);
```

##### ProtectedValue
Used for passwords and custom fields, stored the value in memory XOR'ed  
```ts
let value = new kdbxweb.ProtectedValue(xoredByted, saltBytes);
let valueFromString = kdbxweb.ProtectedValue.fromString('str');
let valueFromBinary = kdbxweb.ProtectedValue.fromBinary(data);
let textString = value.getText();
let binaryData = value.getBinary();
let includesSubString = value.includes('foo');
```

##### Errors
```ts
try {
    await kdbxweb.Kdbx.load(data, credentials);
} catch (e) {
    if (e instanceof kdbxweb.KdbxError && e.code === kdbxweb.Consts.ErrorCodes.BadSignature) {
        /* ... */
    }
}
```

##### Consts
[Consts definition](https://github.com/keeweb/kdbxweb/blob/master/lib/defs/consts.ts)  
```ts
kdbxweb.Consts.ErrorCodes // all thrown errors have code property
kdbxweb.Consts.Defaults // default db settings
kdbxweb.Consts.Icons // icons map
```

##### Random
```ts
let randomArray = kdbxweb.Crypto.random(/* desired length */ 100);
```

##### ByteUtils
```ts
kdbxweb.ByteUtils.bytesToString(bytes);
kdbxweb.ByteUtils.stringToBytes(str);
kdbxweb.ByteUtils.bytesToBase64(bytes);
kdbxweb.ByteUtils.base64ToBytes(str);
kdbxweb.ByteUtils.bytesToHex(bytes);
kdbxweb.ByteUtils.hexToBytes(str);
```

## Building

Use npm to build this project:  
```sh
npm run build
```  

To run tests:  
```sh
npm test
```  

## 3rd party libs

kdbxweb includes these 3rd party libraries:
- [fflate](https://github.com/101arrowz/fflate)
- [xmldom](https://github.com/xmldom/xmldom)

## Tools

The library provides a number of scripts to work with KDBX files:

Dump the binary header:
```sh
npm run script:dump-header my-db.kdbx
```

Print detailed size information about internal objects:
```sh
npm run script:kdbx-size-profiler my-db.kdbx password
```

Dump the internal XML:
```sh
npm run script:kdbx-to-xml my-db.kdbx password
```

Generate big files for load testing:
```sh
npm run script:make-big-files
```

## See it in action

This library is used in [KeeWeb](https://app.keeweb.info)

## Extras

We also provide a template for [HexFiend](https://github.com/ridiculousfish/HexFiend)
to explore the contents of KDBX files, you can find it
[here](format).

## License

MIT


release-notes.md
Release notes
-------------
##### v2.1.1 (2021-09-06)
`-` updated dependencies

##### v2.1.0 (2021-06-17)
`-` replaced `pako` with `fflate` to reduce bundle size  

##### v2.0.5 (2021-05-25)
`-` writable `passwordHash` and `keyFileHash` credentials properties  

##### v2.0.4 (2021-05-19)
`-` fixed saving KDBX3 files without compression  

##### v2.0.3 (2021-05-14)
`-` fixed a crash in Electron when KDBX3 contains large attachments  

##### v2.0.2 (2021-05-10)
`+` exposed `ProtectedValue.salt` and `ProtectedValue.value`  
`+` new static method: `ProtectedValue.fromBase64`
`+` new instance method: `ProtectedValue::toBase64`  

##### v2.0.1 (2021-05-08)
`+` added `passwordHash` and `keyFileHash` to credentials  

##### v2.0.0 (2021-05-08)
`*` the library rewritten in TypeScript  
`*` dist files removed from the repo  
`*` browser support limited to two latest versions  
`*` `Random` removed, use `CryptoEngine.random` instead  
`*` creating `KdbxUuid` with bad values produces an error  
`*` all library modules added to exports  
`*` underscore removed from 'private' class methods  
`*` some object exports replaced with ES6 exports  
`*` binaries interface changed completely, see kdbx-binary.js  
`*` replaced `forEach` with `*allItems` `*allGroups` `*allEntries`    
`*` `object {}` => `Map<>`: `*.customData`, `meta.customIcons`, `entry.fields`, `entry.binaries`  
`*` KDBX 4.1 support  
`*` custom icons contain name and lastModified when possible  
`*` error for files with too high minor version  

##### v1.14.4 (2021-03-19)
`*` cleaning up kdbx.xml after save to save memory

##### v1.14.3 (2021-03-19)
`*` cleaning up kdbx.xml after load to save memory

##### v1.14.2 (2021-02-01)
`-` fixed setting KDF to Argon2id  

##### v1.14.1 (2020-12-31)
`-` fixed new lines removal in non-encoded fields  

##### v1.14.0 (2020-12-30)
`-` fixed parsing xml with bad characters  

##### v1.13.0 (2020-12-09)
`+` possibility to generate a V2 keyfile  
`*` createRandomKeyFile returns a promise  

##### v1.12.1 (2020-12-07)
`+` relaxed keyfile version checking  

##### v1.12.0 (2020-12-07)
`+` V2 keyfiles support  

##### v1.11.0 (2020-12-03)
`+` Argon2id support  

##### v1.10.0 (2020-09-12)
`-` fixed KeyEncryptionRounds header field data type  
`-` missing polyfill added for old Edge versions  

##### v1.9.0 (2020-06-04)
`*` removed text-encoding polyfill  
`+` development: eslint  
`+` development: prettier  

##### v1.8.0 (2020-06-02)
`*` default format changed to KDBX4  

##### v1.7.1 (2020-05-31)
`+` fixed empty icon not understood by other clients  

##### v1.7.0 (2020-05-31)
`+` fixed empty auto-type obfuscation setting not understood by other clients  

##### v1.6.0 (2020-04-10)
`+` challenge-response keys support  

##### v1.5.8 (2020-03-15)
`+` fixed historyMaxItems for 0 and -1  

##### v1.5.7 (2019-12-01)
`+` fixed minor version for v4 files  

##### v1.5.6 (2019-10-26)
`+` generating missing ids while reading files  

##### v1.5.5 (2019-10-04)
`-` fixed another bug in importing entries  

##### v1.5.4 (2019-10-04)
`-` fixed importing entries  

##### v1.5.3 (2019-09-24)
`-` fix #26: library usage issues in node.js  

##### v1.5.2 (2019-09-22)
`-` removed a leaked dependency  

##### v1.5.1 (2019-09-22)
`-` fixed a bug in importing attachments  

##### v1.5.0 (2019-09-22)
`+` importing entries from other files using `Kdbx.importEntry`  
`*` debug and release versions are now provided: kdbxweb.js and kdbxweb.min.js  

##### v1.4.2 (2019-09-14)
`+` default encryption rounds increased to 300000  

##### v1.4.1 (2019-09-14)
`+` setting file KDF with `kdbx.setKdf`  

##### v1.4.0 (2019-09-14)
`+` setting file version with `kdbx.setVersion`  

##### v1.3.0 (2019-09-08)
`-` pretty-printing xml option in `Kdbx::saveXml`  

##### v1.2.7 (2019-03-06)
`-` fixed header after upgrade to kdbx4  

##### v1.2.6 (2018-12-19)
`*` performance improvement  

##### v1.2.5 (2018-11-10)
`+` removed usages of obsolete Buffer() constructor  

##### v1.2.4 (2018-07-13)
`+` fixed large attachments error: keeweb/keeweb#922  

##### v1.2.3 (2018-03-29)
`+` throw an error if there's not enough data in a file  

##### v1.2.2 (2018-03-03)
`+` copyright year updated  

##### v1.2.1 (2018-03-03)
`+` support AES KDF in KDBX4  

##### v1.2.0 (2018-03-03)
`!` dropped IE support  

##### v1.1.0 (2018-02-14)
`+` support ChaCha2 in KDBX3  

##### v1.0.2 (2017-09-29)
`-` improved decoding performance, fix #17  

##### v1.0.1 (2017-02-27)
`-` fix opening db with empty binaries  

##### v1.0.0 (2017-02-01)
`+` KDBX4 support  
`!` API updated  

##### v0.4.6 (2016-08-23)
`-` fix keyfiles with bom  

##### v0.4.5 (2016-08-21)
`+` support raw 32-byte and hex 64-byte keyfiles  

##### v0.4.4 (2016-08-16)
`-` fix keyfiles with unicode characters  

##### v0.4.3 (2016-07-30)
`-` index bugfix for v4.0.2  

##### v0.4.2 (2016-07-30)
`+` target index argument in move function  

##### v0.4.1 (2016-04-21)
`-` fixed bug in Firefox  

##### v0.4.0 (2016-04-21)
`!` xmldom is now external dependency  
`-` updated xmldom to patched version without encoder bug  

##### v0.3.11 (2016-04-10)
`-` create recycle bin if it's enabled but not yet created   

##### v0.3.10 (2016-04-03)
`-` Fixed random keyfile generator  

##### v0.3.8 (2016-03-04)
`+` Expose Kdbx.Consts.Signatures  

##### v0.3.7 (2016-03-04)
`-` Preserve empty fields in entries  

##### v0.3.6 (2016-03-01)
`+` Kdbx.loadXml  

##### v0.3.5 (2016-02-26)
`+` Allow to open db with empty password and keyfile  
`+` Using secure random generator if it's available  

##### v0.3.4 (2016-02-14)
KdbxCredentials.createKeyFileWithHash  

##### v0.3.3 (2015-12-17)
Binaries management  

##### v0.3.2 (2015-12-13)
ASCII-only dist  

##### v0.3.1 (2015-12-02)
Version fix  

##### v0.3.0 (2015-12-02)
Merge  
`+` Kdbx.merge  
`+` Kdbx.[get,set,remove]LocalEditState  
`+` KdbxEntry.removeHistory  
`+` KdbxGroup.forEach now accepts thisArg  

##### v0.2.6 (2015-11-22)
Custom icons cleanup  

##### v0.2.6 (2015-11-21)

lib/defs/consts.ts
export const Signatures = {
    FileMagic: 0x9aa2d903,
    Sig2Kdbx: 0xb54bfb67,
    Sig2Kdb: 0xb54bfb65
} as const;

export const ErrorCodes = {
    NotImplemented: 'NotImplemented',
    InvalidArg: 'InvalidArg',
    BadSignature: 'BadSignature',
    InvalidVersion: 'InvalidVersion',
    Unsupported: 'Unsupported',
    FileCorrupt: 'FileCorrupt',
    InvalidKey: 'InvalidKey',
    MergeError: 'MergeError',
    InvalidState: 'InvalidState'
} as const;

export const CompressionAlgorithm = {
    None: 0,
    GZip: 1
} as const;

export const CrsAlgorithm = {
    Null: 0,
    ArcFourVariant: 1,
    Salsa20: 2,
    ChaCha20: 3
} as const;

export const KdfId = {
    Argon2: '72Nt34wpREuR96mkA+MKDA==',
    Argon2d: '72Nt34wpREuR96mkA+MKDA==',
    Argon2id: 'nimLGVbbR3OyPfw+xvCh5g==',
    Aes: 'ydnzmmKKRGC/dA0IwYpP6g=='
} as const;

export const CipherId = {
    Aes: 'McHy5r9xQ1C+WAUhavxa/w==',
    ChaCha20: '1gOKK4tvTLWlJDOaMdu1mg=='
} as const;

export const AutoTypeObfuscationOptions = {
    None: 0,
    UseClipboard: 1
} as const;

export const Defaults = {
    KeyEncryptionRounds: 300000,
    MntncHistoryDays: 365,
    HistoryMaxItems: 10,
    HistoryMaxSize: 6 * 1024 * 1024,
    RecycleBinName: 'Recycle Bin'
} as const;

export const Icons = {
    Key: 0,
    World: 1,
    Warning: 2,
    NetworkServer: 3,
    MarkedDirectory: 4,
    UserCommunication: 5,
    Parts: 6,
    Notepad: 7,
    WorldSocket: 8,
    Identity: 9,
    PaperReady: 10,
    Digicam: 11,
    IRCommunication: 12,
    MultiKeys: 13,
    Energy: 14,
    Scanner: 15,
    WorldStar: 16,
    CDRom: 17,
    Monitor: 18,
    EMail: 19,
    Configuration: 20,
    ClipboardReady: 21,
    PaperNew: 22,
    Screen: 23,
    EnergyCareful: 24,
    EMailBox: 25,
    Disk: 26,
    Drive: 27,
    PaperQ: 28,
    TerminalEncrypted: 29,
    Console: 30,
    Printer: 31,
    ProgramIcons: 32,
    Run: 33,
    Settings: 34,
    WorldComputer: 35,
    Archive: 36,
    Homebanking: 37,
    DriveWindows: 39,
    Clock: 39,
    EMailSearch: 40,
    PaperFlag: 41,
    Memory: 42,
    TrashBin: 43,
    Note: 44,
    Expired: 45,
    Info: 46,
    Package: 47,
    Folder: 48,
    FolderOpen: 49,
    FolderPackage: 50,
    LockOpen: 51,
    PaperLocked: 52,
    Checked: 53,
    Pen: 54,
    Thumbnail: 55,
    Book: 56,
    List: 57,
    UserKey: 58,
    Tool: 59,
    Home: 60,
    Star: 61,
    Tux: 62,
    Feather: 63,
    Apple: 64,
    Wiki: 65,
    Money: 66,
    Certificate: 67,
    BlackBerry: 68
} as const;
